File: digit_factorial_chains.py
from math import factorial

cache = {}


def factorialSum(number):
    return sum([factorial(int(num)) for num in str(number)])


print(factorialSum(169))


def countRepeatingTerms(number):
    encountered_numbers = [number]

    if number in cache:
        return cache[number]

    while True:
        next_number = factorialSum(encountered_numbers[-1])
        if next_number in encountered_numbers:
            cache[number] = encountered_numbers
            return encountered_numbers
        else:
            encountered_numbers.append(next_number)


def digitFactorialChains():
    count = 0

    for number in range(int(1e3), int(1e6) + 1):
        if countRepeatingTerms(number) == 60:
            count += 1

    return count


# print(digitFactorialChains())

File: digit_factorial_chains.cpp
#include <iostream>
#include <string>
using namespace std;

int factorialSum(int number)
{
    string num_str = to_string(number);
    long factorials[10] = {1, 1, 2, 6, 24, 120, 720, 5040, 40320, 362880};
    long sum = 0;

    for (int i = 0; i < num_str.length(); i++)
    {
        cout << num_str[i] << "\n";
        sum += factorials[num_str[i]];
    }

    return sum;
}

int main()
{
    cout << factorialSum(169) << endl;
    return 0;
}